pageserver/src/billing_metrics.rs

```diff
@@ -0,0 +1,278 @@
+
+    info!("starting collect_metrics");
+
+    // define client here to reuse it for all requests
+    let client = reqwest::Client::new();
+    let cached_metrics: Arc<Mutex<HashMap<BillingMetricsKey, u64>>> =
+        Arc::new(Mutex::new(HashMap::new()));
+
+    loop {
+        tokio::select! {
+            _ = task_mgr::shutdown_watcher() => {
+                info!("collect_metrics received cancellation request");
+                return Ok(());
+            },
+            _ = ticker.tick() => {
+                let cached_metrics = cached_metrics.clone();
+                collect_metrics_task(&client, &cached_metrics, &conf.metric_collection_endpoint).await?;
+            }
+        }
+    }
+}
+
+// One iteration of metrics collection
+// Gather per-tenant and per-timeline metrics and send them to billing service,
+// cache them to avoid sending the same metrics multiple times.
+pub async fn collect_metrics_task(
+    client: &reqwest::Client,
+    cached_metrics: &Arc<Mutex<HashMap<BillingMetricsKey, u64>>>,
+    metric_collection_endpoint: &str,
+) -> anyhow::Result<()> {
+    let mut current_metrics: Vec<(BillingMetricsKey, u64)> = Vec::new();
+    trace!("starting collect_metrics_task");
+
+    // get list of tenants
+    let tenants = tenant_mgr::list_tenants().await;
+
+    // iterate through list of Active tenants and collect metrics
+    for (tenant_id, tenant_state) in tenants {
+        if tenant_state != TenantState::Active {
+            continue;
+        }
+
+        let tenant = tenant_mgr::get_tenant(tenant_id, true).await?;
+
+        let mut tenant_physical_size = 0;
+
+        // iterate through list of timelines in tenant
+        for timeline in tenant.list_timelines().iter() {
+            let timeline_written_size = u64::from(timeline.get_last_record_lsn());
+
+            current_metrics.push((
+                BillingMetricsKey {
+                    tenant_id,
+                    timeline_id: Some(timeline.timeline_id),
+                    metric: BillingMetricKind::WrittenSize.to_string(),
+                },
+                timeline_written_size,
+            ));
+
+            let timeline_size = timeline.get_physical_size();
+            tenant_physical_size += timeline_size;
+
+            debug!(
+                "per-timeline current metrics for tenant: {}: timeline {} physical_size={} last_record_lsn {} (as bytes)",
+                tenant_id, timeline.timeline_id, timeline_size, timeline_written_size)
+        }
+
+        let tenant_remote_size = tenant.get_remote_size().await?;
+        debug!(
+            "collected current metrics for tenant: {}: state={:?} tenant_physical_size={} remote_size={}",
+            tenant_id, tenant_state, tenant_physical_size, tenant_remote_size
+        );
+
+        current_metrics.push((
+            BillingMetricsKey {
+                tenant_id,
+                timeline_id: None,
+                metric: BillingMetricKind::PhysicalSize.to_string(),
```

Comment:
Why don't you use the enum and all the time you're converting the value to string?